configure.ac

```diff
@@ -1500,7 +1500,7 @@ AC_ARG_WITH(pm,
 
 case "$with_pmi" in
   slurm)
-    with_pmi=pmi2
+    with_pmi=pmi1
     with_pmilib=slurm
     with_pm=no
     ;;
```

Comment:
Have you tested PMI1 on Cray? Last time I tried it did not seem to work. This was years ago, so things might have changed.

Reply:
Indeed --
```
Fatal error in internal_Init: Other MPI error, error stack:
internal_Init(45298).....: MPI_Init(argc=0x7fffffff68fc, argv=0x7fffffff68f0) failed
MPII_Init_thread(208)....:
MPID_Init(67)............:
init_world(171)..........: channel initialization failed
MPIDI_CH3_Init(84).......:
MPID_nem_init(395).......:
MPIDI_PG_SetConnInfo(513):
MPIR_pmi_kvs_put(278)....: PMI_KVS_Put returned -1
aborting job:
Fatal error in internal_Init: Other MPI error, error stack:
internal_Init(45298).....: MPI_Init(argc=0x7fffffff68fc, argv=0x7fffffff68f0) failed
MPII_Init_thread(208)....:
MPID_Init(67)............:
init_world(171)..........: channel initialization failed
MPIDI_CH3_Init(84).......:
MPID_nem_init(395).......:
MPIDI_PG_SetConnInfo(513):
MPIR_pmi_kvs_put(278)....: PMI_KVS_Put returned -1

```

Reply:
quote from `cray/pe/pmi/5.0.16/include/pmi2.h`:
```
/* BEGIN CRAY ADDITION */
/*
 * Please note that Cray PMI API and behavior *must* match
 * ANL MPICH PMI for those  parts of PMI and PMI2 that are
 * maintained by ANL.  Consequently this file (pmi2.h) should
 * match the ANL MPICH pmi2.h as much as possible.  Cray
 * additions to the PMI API should be added to the Cray PMI
 * extensions file (pmi_cray_ext.h), not this file.
 *
 * Cray modifications to this file should be denoted similar to
 * how this comment has been denoted.  Following this convention
 * will make it easier to compare this file to the standard ANL
 * MPICH pmi.h file to verify that definitions and function
 * proto-types agree between the Cray PMI and ANL MPICH PMI.
 */

/* Cray modifications:
 * Cray PMI supports a superset of PMI (both PMI-1 and PMI-2)
 * - removed #ifndef USE_PMI2_API
 * - add a USE_PMI2_API check before the definition of PMI_keyval_t
 *   to prevent a redefinition error.  PMI_keyval_t is defined in
 *   both pmi.h and pmi2.h.  In Cray's PMI the definition is removed
 *   from pmi2.h, but this can't be done for MPICH2 which includes
 *   only pmi2.h if USE_PMI2_API is set.
 * - Add code at the end to include the Cray PMI extensions file
 *   (pmi_cray_ext.h) rather than in pmi.h.  See pmi.h comments
 *   for an explanation.
 */
/* END CRAY ADDITION */

/* BEGIN CRAY DELETION */
/*
 * Remove the following check.  Cray includes pmi2.h in pmi.h.
 */
/* #ifndef USE_PMI2_API
 * #error This header file defines the PMI2 API, but PMI2 was not selected
 * #endif
 */
/* END CRAY DELETION */
```

Reply:
With PMI2:
```
Fatal error in internal_Init: Other MPI error, error stack:
internal_Init(45298).: MPI_Init(argc=0x7fffffff68fc, argv=0x7fffffff68f0) failed
MPII_Init_thread(208):
MPID_Init(64)........:
init_local(116)......: channel initialization failed
init_pg(313).........:
MPIR_pmi_init(163)...: PMI2_Job_GetId returned 14
[unset]: aborting job:
Fatal error in internal_Init: Other MPI error, error stack:
internal_Init(45298).: MPI_Init(argc=0x7fffffff68fc, argv=0x7fffffff68f0) failed
MPII_Init_thread(208):
MPID_Init(64)........:
init_local(116)......: channel initialization failed
init_pg(313).........:
MPIR_pmi_init(163)...: PMI2_Job_GetId returned 14
```

Reply:
My build scripts had a mistake that linked in the embedded libpmi.la rather than just `-lpmi`. After fix that, it worked with Cray using PMI 2. PMI 1 with cray is still broken.

Reply:
Added some debug prints for `Cray + PMI1`:
```
thetamom1 /gpfs/mira-home/hzhou/pull_requests/mpich5909$ aprun -n 1 ./cpi
has_parent=0, rank=0, size=1, appnum=0
** PMI_KVS_Put key [P0-businesscard] val [description#nid03834$port#59195$ifname#10.128.15.25$]
Fatal error in internal_Init: Other MPI error, error stack:
internal_Init(45413).....: MPI_Init(argc=0x7fffffff68fc, argv=0x7fffffff68f0) failed
MPII_Init_thread(208)....:
MPID_Init(67)............:
init_world(171)..........: channel initialization failed
MPIDI_CH3_Init(84).......:
MPID_nem_init(395).......:
MPIDI_PG_SetConnInfo(513):
MPIR_pmi_kvs_put(286)....: PMI_KVS_Put returned -1

Application 25758584 exit codes: 15
Application 25758584 resources: utime ~0s, stime ~2s, Rss ~6880, inblocks ~42, outblocks ~0
```
`PMI_Init` succeeded, but `PMI_KVS_Put` failed

Reply:
Asked HPE, PMI1 is not fully supported currently by Cray's `aprun`. In particular, the `kvs-` APIs.